app/discovery/provider/consulcatalog/consulcatalog.go

```diff
@@ -0,0 +1,95 @@
+package consulcatalog
+
+import (
+	"context"
+	"fmt"
+	"github.com/umputun/reproxy/app/discovery"
+	"log"
+	"sync"
+	"time"
+)
+
+//go:generate moq -out consul_client_mock.go -skip-ensure -fmt goimports . ConsulClient
+
+type ConsulClient interface {
+	Get() ([]*ConsulService, error)
+}
+
+type ConsulCatalog struct {
+	client  ConsulClient
+	events  chan discovery.ProviderID
+	autoAPI bool
+
+	listMx sync.RWMutex
+	list   []*ConsulService
+
+	currentServicesID map[string]struct{}
+}
+
+func New(ctx context.Context, client ConsulClient, autoAPI bool, updateInterval time.Duration) *ConsulCatalog {
+	cc := &ConsulCatalog{
+		client:            client,
+		autoAPI:           autoAPI,
+		events:            make(chan discovery.ProviderID),
+		currentServicesID: make(map[string]struct{}),
+	}
+
+	go cc.update(ctx, updateInterval)
+
+	return cc
+}
+
+func (cc *ConsulCatalog) Events(_ context.Context) (res <-chan discovery.ProviderID) {
+	return cc.events
+}
+
+func (cc *ConsulCatalog) update(ctx context.Context, interval time.Duration) {
+	var err error
+	for {
+		err = cc._update()
+		if err != nil {
+			log.Printf("[ERROR] error update consul catalog data")
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(interval):
+		}
+	}
+}
+
+func (cc *ConsulCatalog) _update() error {
```

Comment:
well,  this is not python ;) such _func looks very alien in go. I'd rather put it inside of update above as an internal func